src/resource/resource.c

```diff
@@ -94,9 +100,99 @@ static char* get_name(ResourceHandler *handler, const char *path) {
 	return resource_util_basename(handler->subdir, path);
 }
 
-static Resource* load_resource(ResourceHandler *handler, const char *path, const char *name, ResourceFlags flags) {
+typedef struct ResourceAsyncLoadData {
+	ResourceHandler *handler;
+	char *path;
+	char *name;
+	ResourceFlags flags;
+	void *opaque;
+} ResourceAsyncLoadData;
+
+static int load_resource_async_thread(void *vdata) {
+	ResourceAsyncLoadData *data = vdata;
+	SDL_Event evt;
+
+	data->opaque = data->handler->begin_load(data->path, data->flags);
+
+	SDL_zero(evt);
+	evt.type = sdlevent_finalize_load;
+	evt.user.data1 = data;
+	SDL_PushEvent(&evt);
+
+	log_debug("Thread is exiting");
+	return 0;
+}
+
+static Resource* load_resource_finish(void *opaque, ResourceHandler *handler, const char *path, const char *name, char *allocated_path, char *allocated_name, ResourceFlags flags);
+
+bool resource_sdl_event(SDL_Event *evt) {
+	if(evt->type != sdlevent_finalize_load) {
+		return false;
+	}
+
+	ResourceAsyncLoadData *data = evt->user.data1;
+
+	if(!data) {
+		free(data);
+		return true;
+	}
+
+	char name[strlen(data->name) + 1];
+	strcpy(name, data->name);
+
+	load_resource_finish(data->opaque, data->handler, data->path, data->name, data->path, data->name, data->flags);
+	hashtable_unset(data->handler->async_load_data, name);
+	free(data);
+
+	return true;
+}
+
+static void load_resource_async(ResourceHandler *handler, char *path, char *name, ResourceFlags flags) {
+	log_debug("Loading %s '%s' asynchronously", resource_type_names[handler->type], name);
+
+	ResourceAsyncLoadData *data = malloc(sizeof(ResourceAsyncLoadData));
+	hashtable_set_string(handler->async_load_data, name, data);
+
+	data->handler = handler;
+	data->path = path;
+	data->name = name;
+	data->flags = flags;
+
+	SDL_DetachThread(SDL_CreateThread(load_resource_async_thread, __func__, data));
+}
+
+static void update_async_load_state(void) {
+	SDL_Event evt;
+	while(SDL_PeepEvents(&evt, 1, SDL_GETEVENT, sdlevent_finalize_load, sdlevent_finalize_load)) {
+		resource_sdl_event(&evt);
+	}
+}
+
+static bool resource_check_async_load(ResourceHandler *handler, const char *name) {
+	if(SDL_ThreadID() == main_thread_id)
+		update_async_load_state();
+	ResourceAsyncLoadData *data = hashtable_get_string(handler->async_load_data, name);
+	return data;
+}
+
+static void resource_wait_for_async_load(ResourceHandler *handler, const char *name) {
+	// XXX: is there a better way than a busy loop?
```

Comment:
Wouldn't that just make it less responsive? That comment is kinda old/irrelevant, I actually think a busy loop is ok for such short-ish waits now, but correct me if I'm wrong.